Support private email as uid

bcat-eu opened the pull request:
Handle email uid as a special case, s. https://github.com/ueberauth/ueberauth_github/issues/10#issuecomment-258899587

doomspork commented on lib/ueberauth/strategy/github.ex:
```diff
@@ -129,11 +129,13 @@ defmodule Ueberauth.Strategy.Github do
+    uid_field = conn |> option(:uid_field) |> to_string()
+    if (uid_field == "email") do
+      # private email will not be available as :email and must be fetched
+      conn.private.github_user["email"] || Enum.find(conn.private.github_user["emails"] || [], &(&1["primary"]))["email"]
```
@bcat-eu won't this line throw an error in the event that "emails" is `nil`?  I believe the result code will equate to `[]["email"]` which throws an error.

bcat-eu commented:
@doomspork :+1: to pattern matching change, I thought about it but didn't want to change to much. I believe this `defp fetch_uid("email", %{private: %{github_user: %{"email" => email}}), do: email` will always match (on email nil) but the idea is good and doable.

> won't this line throw an error in the event that "emails" is nil? I believe the result code will equate to []["email"] which throws an error.

it might but it's the same line that we already have [here](https://github.com/ueberauth/ueberauth_github/blob/master/lib/ueberauth/strategy/github.ex#L166). Other change I thought about would be moving the `Enum.find(user["emails"] || [], &(&1["primary"]))["email"]` part to a `get_primary_email` function, we could refactor it there. Would it make sense?

As for the error itself, my understanding was that we can not proceed without an email so an error sounds good, a more explicit exception might make it friendlier.

bcat-eu commented:
@doomspork I've made some adjustments as described above:

* moved email fetching to a function and called it from both places (uid and info) where the `Enum.find(user["emails"] || [], &(&1["primary"]))["email"]` was used
* refactored the email fetching logic a little, dropped the fallback `[]` since it would lead to `[]["email"]` and added an explicit exception for the case where we don't get any emails (assuming GitHub never sends emails with no primary address). This should never happen so the exception should be fine, in case of this PR (user wants email as UID) it's 100% fine
* split the uid fetching into couple functions to replace `if` with pattern matching and smaller functions

doomspork commented:
Howdy @bcat-eu!  Sorry for the delay, I was away for awhile on vacation and I'm now getting back to all of these 👍

doomspork commented on lib/ueberauth/strategy/github.ex:
```diff
@@ -163,7 +159,7 @@ defmodule Ueberauth.Strategy.Github do
       name: user["name"],
       nickname: user["login"],
-      email: user["email"] || Enum.find(user["emails"] || [], &(&1["primary"]))["email"],
+      email: user["email"] || get_primary_email!(user),
```
Should we create a `fetch_email` function for this code?  I noticed it's re-used in the `fetch_uid/2` function.

```elixir
defp fetch_email(user), do: user["email"] || get_primary_email!(user)
```

doomspork commented on lib/ueberauth/strategy/github.ex:
```diff
@@ -195,6 +191,23 @@ defmodule Ueberauth.Strategy.Github do
     }
   end
 
+  defp fetch_uid("email",  %{private: %{github_user: user}}) do
```
There's two spaces after the comma, could you remove the extra one?

bcat-eu commented:
> sorry for the delay! Looking good, two small pieces of feedback then I will get this merged and a release cut

you got it :) One thing though: I know it compiles but I didn't push it to server for testing after this last change and there are no tests so we just know it must work. I can't see what could break here though.

bcat-eu commented:
@doomspork bump? :)

doomspork commented:
Thank you @bcat-eu, sorry for the delay!  I will cut a release tomorrow, I'm quite busy today 👍
